docs/intro/fastapi_gelai_searchbot.rst

```diff
@@ -0,0 +1,1556 @@
+
+    $ python3 app/web.py
+
+
+It's time to reflect the new capabilities in our web app. We're going to hack
+our way around the fact that many stories on HN come with no comment threads.
+
+.. code-block:: python
+     :caption: app/main.py
+
+     from .web import fetch_web_sources, WebSource
+
+     async def search_web(query: str) -> list[WebSource]:
+         raw_souces = fetch_web_sources(query, limit=30)
+         return [s for s in raw_souces if s.text is not None][:5]
+
+
+Now we can update the ``/search`` endpoint as follows:
+
+.. code-block:: python-diff
+    :caption: app/main.py
+
+      class SearchResult(BaseModel):
+          response: str | None = None
+    +     sources: list[WebSource] | None = None
+
+
+      @app.post("/search")
+      async def search(search_terms: SearchTerms) -> SearchResult:
+    +     web_sources = await search_web(search_terms.query)
+    -     return SearchResult(response=search_terms.query)
+    +     return SearchResult(
+    +         response=search_terms.query, sources=web_sources
+    +     )
+
+
+4. Connect to the LLM
+=====================
+
+Now that we're capable of scraping text from search results, we can forward
+those results to the LLM to get a nice-looking summary.
+
+There's a million different LLMs accessible via a web API, feel free to choose
+whichever you prefer. In this tutorial we will roll with OpenAI, primarily for
+how ubiquitous it is. To keep things somewhat provider-agnostic, we're going to
+get completions via raw HTTP requests. Let's grab API descriptions from
+OpenAI's `API documentation
+<https://platform.openai.com/docs/api-reference/chat/create>`_, and set up LLM
+generation like this:
+
+.. code-block:: python
+    :caption: app/main.py
+
+    import requests
+    from dotenv import load_dotenv
+
+    _ = load_dotenv()
+
+
+    def get_llm_completion(system_prompt: str, messages: list[dict[str, str]]) -> str:
+        api_key = os.getenv("OPENAI_API_KEY")
+        url = "https://api.openai.com/v1/chat/completions"
+        headers = {"Content-Type": "application/json", "Authorization": f"Bearer {api_key}"}
+
+        response = requests.post(
+            url,
+            headers=headers,
+            json={
+                "model": "gpt-4o-mini",
+                "messages": [
+                    {"role": "developer", "content": system_prompt},
+                    *messages,
+                ],
+            },
+        )
+        response.raise_for_status()
+        result = response.json()
+        return result["choices"][0]["message"]["content"]
+
+
+Note that this cloud LLM API (and many others) requires a secret key to be set
+as an environment variable. A common way to manage those is to use the
+``python-dotenv`` library in combinations with a ``.env`` file. Feel free to
+browse `the readme
+<https://github.com/theskumar/python-dotenv?tab=readme-ov-file#getting-started>`_,
+although it's also quite simple to use. Create a file called ``.env`` in the
```

Comment:
(So that we don't assume something is simple)
```suggestion
``python-dotenv`` library in combinations with a ``.env`` file. Feel free to
browse `the readme
<https://github.com/theskumar/python-dotenv?tab=readme-ov-file#getting-started>`_,
to learn more. Create a file called ``.env`` in the
```